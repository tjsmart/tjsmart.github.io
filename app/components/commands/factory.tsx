type CommandCallback = (args: string[]) => string | JSX.Element;

export class Command {
    public name: string;
    public callback: CommandCallback;
    public description: string | null = null;

    public constructor(
        name: string,
        callback: CommandCallback,
        description: string | null = null,
    ) {
        this.name = name;
        this.callback = callback;
        this.description = description;
    }
}

const commandMap = new Map<string, Command>();

/**
 * Register a command to the factory.
 *
 * @param command - the command to be registered
 */
export function registerCommand(command: Command) {
    // console.log(commandMap);
    if (commandMap.has(command.name)) {
        throw new Error(`Command ${command.name} is already registered!`);
    }
    commandMap.set(command.name, command);
}

/**
 * Handles the provided command line.
 *
 * @param commandLine - the input text to be handled
 * @returns The output from the command (if any)
 */
export function handleCommand(commandLine: string): JSX.Element {
    // TODO: better parsing...
    let [name, ...args] = commandLine.trim().split(" ");

    const command = commandMap.get(name);
    if (!command) {
        return <span> {`command not found: ${name}`} </span>;
    }

    const output = command.callback(args);
    if (typeof output === "string") {
        return (
            <>
                {output.split("\n").map((line: string) => (
                    <>
                        {line}
                        <br />
                    </>
                ))}
            </>
        );
    }
    return output;
}

const lb =
    "--------------------------------------------------------------------------------";

const help = new Command(
    "help",
    (args: string[]): string => {
        if (args.length == 0) {
            let outputLines = new Array("Available Commands:", lb);
            let commandDetails = Array.from(commandMap.values()).map(
                (value: Command) => `${value.name} - ${value.description}`,
            );

            outputLines.push(...commandDetails);
            return outputLines.join("\n");
        }
        if (args.length > 1) {
            // TODO: error
            return "Too many arguments";
        }
        let command = commandMap.get(args[0]);
        if (!command) {
            // TODO: error
            return `command not found: ${args[0]}`;
        }
        if (!command.description) {
            // TODO: error
            return `no help found for: ${command.name}`;
        }
        // TODO: formatting?
        // TODO: should length of line be the length of 'Help on command.name'? Or maybe length of screen?
        return `Help on \`${command.name}\`\n${lb}\n${command.description}`;
    },
    "Display help on the provided command, e.g. `help hello`",
);

registerCommand(help);

if (process.env["NODE_DEV"] == "TEST") {
    module.exports.help = help;
    module.exports.commandMap = commandMap;
}
