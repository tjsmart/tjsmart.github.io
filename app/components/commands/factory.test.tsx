import renderer from "react-test-renderer";
import { registerCommand, handleCommand, Command } from "./factory";
import * as factory from "./factory";

// @ts-ignore
const help: Command = factory.help;

const test_cmd = new Command(
    "test_cmd",
    (args: string[]) => `test command!${args.join(" ")}`,
    "Just a test command.",
);

let storedCommandMap: Map<string, Command>;

beforeEach(() => {
    // @ts-ignore
    storedCommandMap = factory.commandMap;
    // @ts-ignore
    factory.commandMap.clear();
    registerCommand(test_cmd);
});
// TODO: set commandMap state back
// afterAll(() => {
//     // @ts-ignore
//     factory.commandMap = storedCommandMap;
// });

describe("factory", () => {
    test("registered command executes expected callback", () => {
        let result = handleCommand("test_cmd");
        let tree = renderer.create(result).toJSON();
        expect(tree).toMatchSnapshot();

        result = handleCommand("test_cmd foo bar baz");
        tree = renderer.create(result).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("handle unknown command output", () => {
        const result = handleCommand("i_am_unknown");
        const tree = renderer.create(result).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("handle command with extra whitespace", () => {
        const getJSON = (cmd: string) =>
            renderer.create(handleCommand(cmd)).toJSON();

        const actual = getJSON("   test_cmd   ");
        const expected = getJSON("test_cmd");

        expect(actual).toEqual(expected);
    });

    test("cannot register a command twice", () => {
        expect(() => registerCommand(test_cmd)).toThrowError();
    });

    test("help without argument displays available commands", () => {
        const output = help.callback([]);

        if (typeof output != "string") {
            fail("Expected output to be of type string");
        }

        expect(output.startsWith("Available Commands")).toBe(true);
        expect(output.includes("test_cmd")).toBe(true);
    });

    test("help with valid command, shows help on that command", () => {
        const output = help.callback(["test_cmd"]);

        if (typeof output != "string") {
            fail("Expected output to be of type string");
        }

        expect(output.startsWith("Help on `test_cmd`")).toBe(true);
        expect(output.includes("Just a test command.")).toBe(true);
    });

    test("help with two arguments throws", () => {
        let output = help.callback(["arg1", "arg2"]);

        if (typeof output != "string") {
            fail("Expected output to be of type string");
        }

        expect(output.startsWith("Too many arg")).toBe(true);
    });

    test("help on command without a description", () => {
        test_cmd.description = null;
        let output = help.callback(["test_cmd"]);

        if (typeof output != "string") {
            fail("Expected output to be of type string");
        }

        expect(output.startsWith("no help found")).toBe(true);
        expect(output.includes("test_cmd")).toBe(true);
    });
});
