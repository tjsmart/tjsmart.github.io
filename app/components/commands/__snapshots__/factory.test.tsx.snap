// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory handle unknown command output 1`] = `
<span>
   
  command not found: i_am_unknown
   
</span>
`;

exports[`factory registered command executes expected callback 1`] = `
<div>
  test command!
  <br />
</div>
`;

exports[`factory registered command executes expected callback 2`] = `
<div>
  test command!foo bar baz
  <br />
</div>
`;
